### PR TITLE
NPCs assist with batch crafting & construction...

### PR DESCRIPTION
**Pull request by acidia**
Pandromidal made a request: http://smf.cataclysmdda.com/index.php?topic=9977.msg229036#msg229036

If NPCs have the basic skill to craft an item they will attempt to help.  Most recipies still don't have a batch craft bonus so that really needs to be addressed, even a 5% bonus will allow the NPC effect to kick in.  I would have made it apply for all batches but I figured somethings should only be done by a single person.  In implementation, 1 skilled NPC drops batch time to 75% and 2 drops it to 60%... never to go below the time it takes to craft a single unit.  

alone -> 5 clean water in 10min
1 npc -> 8 in 10 min
2 npc -> 11 in 10 min

The NPC will either 'watch you' meaning they are gaining a bit of exp but not skilled enough to help.  'Assists you' meaning that you created a batch they were skilled enough to help with.  'Could help you with a batch'... what it says. 

I'll take a look at help with constructions/mechanics later...

### CODE REVIEW

File: src/crafting.cpp [hunk 1]
```diff
@@ -1216,6 +1216,17 @@ int recipe::batch_time(int batch) const
         return time * batch;
     }
 
+    // NPCs around you should assist in batch production if they have the skills
+    int assistants = 0;
+    if (!g->active_npc.empty()) {
```
Comment (BevapDin) [inline src/crafting.cpp]: You really don't need this kind of test, the loop will work properly even if `active_npc` is empty.

**Comment by acidia** [comment]
Doing the same with construction is easy, any thoughts on the bonus additional skilled hands should give?  I was thinking 20% time discount per NPC... capping at 60%?

**Comment by Rivet-the-Zombie** [comment]
I think construction should probably get an even greater reduction than crafting, as in my experience it's far more useful to have helpers when building something big like a structure than it is to have extra helpers when doing something small like cooking.

**Comment by acidia** [comment]
Ok, for now I went with a 75% time reduction per NPC added... since we aren't really checking for additional sets of tools I figured that the quickest you could get something done is 40% of the original time.  Hopefully I can get a chance to look into mechanics and bionic installation tomorrow but this should be good to go... 

Keeping NPCs around might be useful for more than defense.  :)

Wait, what if you had an NPC read books to your character if you were illiterate... that might be an silly gimmick.

**Comment by Rivet-the-Zombie** [comment]
>Wait, what if you had an NPC read books to your character if you were illiterate... that might be an silly gimmick. 

That would actually be pretty awesome. It'd make taking 'Illiterate' a viable choice for once.

**Comment by vache** [comment]
Yeah, illiterate is pretty crippling, and having NPCs help out might make it slightly more worth it.  Plus that's probably a pretty good approximation for how illiterate people get along in real life.

**Comment by Zireael07** [comment]
I like the NPCs reading books idea for another PR ;)

**Comment by Rivet-the-Zombie** [comment]
>I like the NPCs reading books idea for another PR ;)

This right here. It's a great idea, but it should probably go in its own PR.

**Comment by DavidKeaton** [comment]
I would love to nuzzle into NPC, but I'm back onto the item change again. :S

**Comment by KA101** [comment]
On construction, I'd say -1 to skill required per friendly/faction NPC onsite, max reduction of half the  original requirement (round down: with four NPCs on hand, the well, normally requiring 9 skill IIRC, could be dug with skill 5).  I'm not averse to the time bonus either.

We balanced most of the high-end projects around the idea that the PC is doing this "needs workers in about three or four spots" job by xemself, so actually doing them properly required either supreme skill...or actually having the help.

Prepared to queue this now; the skill bonus/reduction/wev can also go in later.

**Comment by vache** [comment]
I think skill reduction might be pushing it, but letting an NPC with high construction skill act as a foreman to unlock the higher level recipes while nearby would be a good idea.  It might be handled the same way as crafting with a book in proximity.  I think a little more unification between construction and crafting would be a good idea first though.

**Comment by KA101** [comment]
Uh, vache, perhaps the reduction could be less dramatic, but it should definitely happen.  As I said, we balanced the skill requirements for doing the projects solo, and I for one have repeatedly noted that a group could do them much more easily.  (One need not be a world-class miner to dig a well, if one has staff on hand to handle the supplies topside and haul you out at the end.)

Just doesn't have to happen in this PR.  ;-)

**Comment by acidia** [comment]
Cool, I'll definitely put in a new PR the batch crafting so that NPCs improve the tasks that typically don't receive a batch bonus.  I'm cool with whatever on the construction side.

What are people's opinions on mechanics?  Time really isn't a handicap since most jobs seem to be done fairly quickly.  The only thing I could really think of is having the difficulty check vs the highest mechanics skill between you and your NPCs.

For bionics, there are a couple of ways we could go about it... since it should be much easier to perform surgery on someone else than it is on yourself.  
1)  The mechanics, electronics, and first aid skills only use the highest of each out of your group.  So an NPC mechanic, electrician, and medic would be ideal.
2)  Whoever has the highest percentage chance to succeed performs the surgery.  One surgeon NPC.
3)  NPCs don't help with the surgery, they negate the failures.  Makes installing less risky but doesn't make it a guaranteed success either.  A bunch of mooks who stand around and stop you from hurting yourself...usually.
4)  Something else...

**Comment by KA101** [comment]
Bionics, I'd say that self-install should ideally get scrapped once the system is workable, so I'd say 1 or 2 with the provision that whoever's receiving the aug can't make roll(s).

**Comment by vache** [comment]
True, but I just feel that 5 guys who know construction but don't know how to dig a well equal out to one guy who knows how to dig a well.  There is something to be said for getting things done as a team that you couldn't do on your own though.  Definitely would have to play with a while and see how it feels to get things right.

As for mechanics, I have no real contribution there.  Most mechanical work is done one person at a time unless it is basic procedural stuff designed to get one thing done quick (think express oil change).  I'd say it probably wouldn't help to have NPCs around unless you had some way to direct them to install a part at one end of a vehicle while you installed a part at the other.  At the very least though I'd think NPCs should be capable of repairing a vehicle, given the skill, tools, and materials necessary to do so.

For bionics, the single surgeon NPC makes the most sense to me.

**Comment by Rivet-the-Zombie** [comment]
>self-install should ideally get scrapped once the system is workable

I disagree. There's no reason to block people from being able to get bionics because they play with NPCs turned off, or because they can't manage to keep a suicidal pink **@** alive for long enough to do the surgery.

I'm totally down with having NPC buddies making the installation easier/safer, but not with making NPC assistance a requirement.

**Comment by KA101** [comment]
> NPCs turned off, or because they can't manage to keep a suicidal pink @ alive for long enough to do the surgery.

Without workarounds for these problems (reprogram dissector/hospital/industrial machinery, for example?) and/or un-suicidalizing the NPCs (or at least making sure there are static ones capable of assisting), the system is not workable, and self-install is therefore not scrapped.

Basically, I agree, and didn't make that clear enough.

**Pull request merged**